### ISSUE TITLE
Shader Editor: #pragma disable_preprocessor Offset error highlight

### ISSUE THREAD

**Issue by haruyou27** [godotengine/godot#79976]
### Godot version

v4.1.1.stable.arch_linux

### System information

OpenGL API 4.2 (Core Profile) Mesa 23.1.4 - Compatibility - Using Device: Intel - Mesa Intel(R) HD Graphics 2500 (IVB GT1)  

### Issue description

This issue should have been fixed as #76479  merged.
However, I can still reproduce this in `v4.1.1.stable.arch_linux` 
![20230728_10h06m09s_grim](https://github.com/godotengine/godot/assets/68549456/88449906-8609-4cb4-8cf3-94160bac0680)

For more info: #76116

**Comment by AThousandShips** [comment]
Please do not remove the headers from the template

**Comment by clayjohn** [comment]
CC @bitsawer

**Comment by bitsawer** [comment]
I don't think this is a bug and it should do the same if you have `disable_preprocessor` enabled or not. Shader language statements are delimited by `;` not lines, so the error seems to show a correct line. Consider how the compiler sees this, the missing ; after `speed` is not an immediate syntax error but perfectly valid code. For example, this is one valid version of similar code with line numbers:

```
1) uniform float speed 
2) = 
3) 0.1;
```

If you would accidentally typo 0.1 like this:

```
1) uniform float speed 
2) = 
3) 0,1;
```

which error line number would you expect to see? I think most would like to see error highlight on line 3.

**Comment by clayjohn** [comment]
That's a good observation. Also, removing ``#pragma disable_preprocessor`` has no effect in this case

**Comment by haruyou27** [comment]
I see, it's just confused me because before your PR is merged (v4.0.2.stable.official),  the highlight is on the exact line where `;` is mising.
![image](https://user-images.githubusercontent.com/68549456/232272893-5d218db4-30a8-4c40-94c2-6c0cdce24766.png)

**Comment by bitsawer** [comment]
@haruyou27 That's how it is in current master too. That's because in that new example that line actually contains an error (`errorLine` is not a valid identifier), so it shows the first error it can, it doesn't have to go any further. Change it to this (add `errorLine` declaration before using it) and see how the error and line number changes, but in both examples the line numbers and errors are correct considering the parsing context:

```glsl
void fragment() {
    float some_num = 72.7;
    float errorLine;
    errorLine
} // <- now error will point here, which is also correct.
```

There is always some tug between parser simplicity and showing good error messages. We could try to guess the intent in cases like this and juggle the line numbers around, but I don't think it's worth the increased complexity. There have also been a few other tweaks in error line numbering, but they exact "guilty" error line can be pretty arbitrary in some cases like this.

If there are no other problems here I'll close this issue, improvements or tweaks to shader language are best discussed in proposals: https://github.com/godotengine/godot-proposals/issues or some other place.
